### DISCUSSION TITLE
Alternate wire and graph paradigms

### DISCUSSION THREAD

**Discussion by cdriesler** [cdriesler/glasshopper.io#85] [Ideas & Requests]
As an educational execise, NodePen has always been about evoking the Grasshopper experience as faithfully as possible. As a brand new web application in the year 2021, it has the potential to reconsider all of these things.

@jmanooch reached out [via twitter](https://twitter.com/jmanooch/status/1420077081879359488?s=20) once and placed himself squarely in camp #2. Will keep running thoughts about new graph expressions / management strategies here.

**Comment by jmanooch** [comment]
right on

I find it Grasshopper so bizarre (as an interface, not as a product) and the technical arguments in favour of it seem to be incoherent, taken together

- it's a 'visual paradigm' .... used to express logic in an visually incoherent way (given that you can, because of how connectors work, put items in a 'flow' but out of left-right visual sequence?!

- it's a 'logic paradigm' ... that makes it unnaturally hard to do very fundamental things in computational logic, specifically 1) execute conditions, 2) execute loops, 3) store data generated during 1 and 2. 

Aside from that, it's fugly as fug. What? For the love of god, why - if, after all, it's a visual paradigm?

Aside from that, it's extremely antagonistic to anything other than seasoned users. The way that the topbar menus, and contextual menus, work is so bizarre. Some kind of logical ordering, and hiearchy/priority of dominant compents, etc, would be so very natural and helpful. Let alone /VISUAL TYPING OF FUCKING ELEMENTS/. Jfc. 

I mean I can't really see /any/ reason why 'visual coding' doesn't start with the vexcode paradigm https://vr.vex.com/ or MIT Scratch https://scratch.mit.edu/projects/editor/?tutorial=getStarted or similar. 

- visual typing
- visual logic
- visual nesting/iteration/conditional
etc

How is anything a 'visual coding' paradigm without that kind of thing? I cannot make it make sense. 

I will also note the scurrilous irony, which I haven't shared before, that - among the most impressive humans I know of, and THE most impressive Heumann - @andrewheumann shares his great love for Gh and its visual coding paradigm ... and expresses that love in part by ... writing a much better interface that sits on top of it and obscures it?!

https://www.food4rhino.com/en/app/human-ui

Anyway, I love it! The /personal/ argument for it is easy: it feels like using a sewing machine in an era of fast fashion, or doing a jigsaw in the era of app-based games. You don't get what you want faster, and it's notable by what it can't do, but it teaches you about yourself and asks you what you are really trying to do.

 But I wish we had options .... 'coding interface as meditation' is ... a slightly odd choice.
